### PR TITLE
Converted HTTP API to Tasks/Events

### PR DESCRIPTION
**Pull request by justinboswell**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-crt-http/HttpConnection.cs [hunk 1]
```diff
@@ -28,39 +30,135 @@ public enum OutgoingBodyStreamState
...
+        {
+            ErrorCode = errorCode;
+        }
+    }
+
+    public abstract class HttpClientStreamEventArgs : EventArgs
+    {
+        public HttpClientStream Stream { get; private set; }
+
+        public HttpClientStreamEventArgs(HttpClientStream stream)
```
Comment (JonathanHenson) [inline aws-crt-http/HttpConnection.cs]: if the setter is private, should the constructor be marked internal?

File: aws-crt-http/HttpConnection.cs [hunk 2]
```diff
@@ -28,39 +30,135 @@ public enum OutgoingBodyStreamState
...
+            if (IncomingHeaders == null)
+                throw new ArgumentNullException("IncomingHeaders");
+            if (StreamComplete == null)
+                throw new ArgumentNullException("StreamComplete");
+        }
+
+        internal OutgoingBodyStreamState OnStreamOutgoingBody(HttpClientStream stream, byte[] buffer, out UInt64 bytesWritten)
+        {
+            var e = new StreamOutgoingBodyEventArgs(stream, buffer);
+            StreamOutgoingBody?.Invoke(stream, e);
```
Comment (JonathanHenson) [inline aws-crt-http/HttpConnection.cs]: Aint C# 4 grand?

File: aws-crt-http/HttpConnection.cs (hunk 1 above)
Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: Good call, fixed.

File: aws-crt-http/HttpConnection.cs (hunk 2 above)
Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: It's glorious. I spent last night looking at some old C# code of mine, thinking about how much less of it there would be now.

File: aws-crt-http/HttpConnection.cs [hunk 3]
```diff
@@ -28,39 +30,135 @@ public enum OutgoingBodyStreamState
...
+            : base(stream)
+        {
+            HasBody = hasBody;
+        }
+    }
+
+    public class IncomingBodyEventArgs : HttpClientStreamEventArgs
+    {
+        public byte[] Data { get; private set; }
+        public UInt64 WindowSize { get; private set; }
```
Comment (graebm) [inline aws-crt-http/HttpConnection.cs]: maybe not worth exposing WindowSize as a read-only thing?
It's always equal to Data.length...
On the other hand it's just one more variable on the class, user's don't need to look at it

Maybe name it WindowDecrementSize? WindowShrinkSize? WindowLoss?

Comment (justinboswell) [reply aws-crt-http/HttpConnection.cs]: That should have been writable publicly, fixed. Good catch!

**Pull request merged**